File: leet_BS_0779_kth_symbol___btree.rs
impl Solution {
    pub fn kth_grammar(n: i32, k: i32) -> i32 {
        let mut L = 1;
        let mut R = 2_i32.pow((n - 1) as u32);
        let mut res = 0;
        for _ in 0..(n - 1) {
            let mid = (R - L) / 2 + L;
            if mid < k {
                L = mid + 1;
                res ^= 1;
            } else {
                R = mid;
            }
        }
        res
    }
}

File: leet_BS_0779_kth_symbol___btree.cc
class Solution {
public:
    int kthGrammar(int n, int k)
    {
        long long   L = 1, R = 1;
        int res = 0;
        int i = -1;
        while (++i < n - 1)
            R *= 2;
        i = -1;
        while (++i < n - 1)
        {
            int mid = (R - L) / 2 + L;
            if (mid < k)
            {
                L = mid + 1;
                res ^= 1;
            }
            else
            {
                R = mid;
            }
        }
        return res;
        
    }
};
